lib/oauth/multi.go

```diff
@@ -0,0 +1,182 @@
+package oauth
+
+import (
+	"context"
+	"encoding/gob"
+	"errors"
+	"fmt"
+	"github.com/enfabrica/enkit/lib/khttp"
+	"github.com/enfabrica/enkit/lib/khttp/kcookie"
+	"github.com/enfabrica/enkit/lib/multierror"
+	"github.com/enfabrica/enkit/lib/token"
+	"math/rand"
+	"net/http"
+	"reflect"
+)
+
+const requiredKey = "required"
+
+type MultiOauth struct {
+	RequiredAuth *Authenticator
+	OptAuth      map[string]*Authenticator
```

Comment:
make OptAuth a simple array: `OptAuth []*Authenticator`. So we have extra authenticator 0, extra authenticator 1, ...

Then, change `CompletedFlows`. Instead of having a `[]string`, make it a simple `int`. Which basically gets incremented, 0, 1, 2, 3, ...

Reasons:
1. Less state to move around (literally, fewer bytes).
2. Authentications all happen in the same order. a) Imagine users having to some times to authenticate with github first, and google oauth next, and some times in the reverse order. b) Imagine having to debug an issue related to how oauth providers work, and only happening in some cases, depending on order.
3. Simpler code - you can remove the uuid generator?